### PR TITLE
[c++ grpc] Simplify thread pool concept

### PR DESCRIPTION
**Pull request by ara-ayvazyan**
This change removes the unnecessary need for `TThreadPool` type parameter and instead uses a more simple runtime `std::function` to express a scheduler concept.

Built on top #884.

### CODE REVIEW

File: cpp/inc/bond/ext/grpc/thread_pool.h [hunk 1]
```diff
@@ -62,7 +44,7 @@ class thread_pool : private boost::asio::io_service
     /// will be created.
     explicit
     thread_pool(size_t numThreads = USE_HARDWARE_CONC)
-        : _work(*this)
+        : _service{ std::make_shared<service>() }
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/thread_pool.h]: Let's split this into two ctor and get rid of the `USE_HARDWARE_CONC` constant completely.

Comment (chwarr) [inline cpp/inc/bond/ext/grpc/thread_pool.h]: As we discussed, so we want to throw if `std::thread::hardware_concurrency()` returns 0?

File: compiler/src/Language/Bond/Codegen/Cpp/Grpc_h.hs [hunk 2]
```diff
@@ -95,14 +95,13 @@ grpc_h export_attribute cpp file imports declarations = ("_grpc.h", [lt|
...
 
-    template <typename TThreadPool>
     class #{proxyName}
     {
     public:
         #{proxyName}(
             const std::shared_ptr< ::grpc::ChannelInterface>& channel,
             std::shared_ptr< ::bond::ext::gRPC::io_manager> ioManager,
-            std::shared_ptr<TThreadPool> threadPool);
+            const ::bond::ext::gRPC::Scheduler& scheduler = {});
```
Comment (ara-ayvazyan) [inline compiler/src/Language/Bond/Codegen/Cpp/Grpc_h.hs]: Note the `= {}` part.

File: cpp/inc/bond/ext/grpc/thread_pool.h [hunk 3]
```diff
@@ -28,87 +30,69 @@
...
-                const size_t recourseNumThreads = 2;
-                numThreads = recourseNumThreads;
-            }
+            throw InvalidThreadCount{};
         }
 
-        // Spin working threads.
-        for (size_t i = 0; i < numThreads; ++i)
+        auto& service = *_service;
+        for (unsigned long i = 0; i < numThreads; ++i)
```
[the remark this answers is not in the document]
Comment (ara-ayvazyan) [reply cpp/inc/bond/ext/grpc/thread_pool.h]: This meant to be [signed] `long` for the [unlikely] counter overflow.

**Pull request merged**